Automated Smoketest 
Sha: ec689b70291d4d15eb506848a330262751b6a1e7
Build: failed
Number of Error(s): 6
  Policy CMP0046 is not set: Error on non-existent dependency in
  Policy CMP0046 is not set: Error on non-existent dependency in
/mnt/disk1/home/vamosax/smoketest/PR-8806/HPCC-Platform/plugins/javaembed/javaembed.cpp:2911:137: error: cannot allocate an object of abstract type \xe2\x80\x98javaembed::JavaXmlBuilder\xe2\x80\x99
make[2]: *** [plugins/javaembed/CMakeFiles/javaembed.dir/javaembed.cpp.o] Error 1
make[1]: *** [plugins/javaembed/CMakeFiles/javaembed.dir/all] Error 2
make: *** [all] Error 2

--- a/common/thorhelper/thorxmlwrite.hpp
+++ b/common/thorhelper/thorxmlwrite.hpp
@@ -42,8 +42,26 @@ interface IXmlWriterExt : extends IXmlWriter
     virtual IXmlWriterExt & clear() = 0;
     virtual size32_t length() const = 0;
     virtual const char *str() const = 0;
-    virtual void rewindTo(unsigned int prevlen) = 0;
+    virtual IInterface *saveLocation() const = 0;
+    virtual void rewindTo(IInterface *location) = 0;
     virtual void outputNumericString(const char *field, const char *fieldname) = 0;
+};
+
+class thorhelper_decl CommonXmlPosition : public CInterface, implements IInterface
+{
+public:
+    IMPLEMENT_IINTERFACE;
+
+    CommonXmlPosition(size32_t _pos, unsigned _indent, unsigned _nestLimit, bool _tagClosed, bool _needDelimiter) :
+        pos(_pos), indent(_indent), nestLimit(_nestLimit), tagClosed(_tagClosed), needDelimiter(_needDelimiter)
+    {}
+
+public:
+    size32_t pos = 0;
+    unsigned indent = 0;
+    unsigned nestLimit = 0;
+    bool tagClosed = false;
+    bool needDelimiter = false;
 };
 
 class thorhelper_decl CommonXmlWriter : public CInterface, implements IXmlWriterExt
@@ -82,12 +100,28 @@ public:
     virtual IXmlWriterExt & clear();
     virtual unsigned length() const                                 { return out.length(); }
     virtual const char * str() const                                { return out.str(); }
-    virtual void rewindTo(unsigned int prevlen)
+    virtual IInterface *saveLocation() const
     {
         if (flusher)
             throwUnexpected();
 
-        if (prevlen < out.length()) out.setLength(prevlen);
+        return new CommonXmlPosition(length(), indent, nestLimit, tagClosed, false);
+    }
+    virtual void rewindTo(IInterface *saved)
+    {
+        if (flusher)
+            throwUnexpected();
+
+        CommonXmlPosition *position = dynamic_cast<CommonXmlPosition *>(saved);
+        if (!position)
+            return;
+        if (position->pos < out.length())
+        {
+            out.setLength(position->pos);
+            tagClosed = position->tagClosed;
+            indent = position->indent;
+            nestLimit = position->nestLimit;
+        }
     }
 
     virtual void outputNumericString(const char *field, const char *fieldname)
@@ -156,6 +190,29 @@ public:
     virtual unsigned length() const                                 { return out.length(); }
     virtual const char * str() const                                { return out.str(); }
     virtual void rewindTo(unsigned int prevlen)                     { if (prevlen < out.length()) out.setLength(prevlen); }
+    virtual IInterface *saveLocation() const
+    {
+        if (flusher)
+            throwUnexpected();
+
+        return new CommonXmlPosition(length(), indent, nestLimit, false, needDelimiter);
+    }
+    virtual void rewindTo(IInterface *saved)
+    {
+        if (flusher)
+            throwUnexpected();
+
+        CommonXmlPosition *position = dynamic_cast<CommonXmlPosition *>(saved);
+        if (!position)
+            return;
+        if (position->pos < out.length())
+        {
+            out.setLength(position->pos);
+            needDelimiter = position->needDelimiter;
+            indent = position->indent;
+            nestLimit = position->nestLimit;
+        }
+    }
 
     void outputBeginRoot(){out.append('{');}
     void outputEndRoot(){out.append('}');}

--- a/esp/esdllib/esdl_transformer2.cpp
+++ b/esp/esdllib/esdl_transformer2.cpp
@@ -703,7 +703,7 @@ void Esdl2Array::process(Esdl2TransformerContext &ctx, IPropertyTree *pt, const 
         const char *tagname = queryOutputName(ctx);
         if (pt->hasChildren())
         {
-            int prevlen = ctx.writer->length();
+            Owned<IInterface> prevLocation = ctx.writer->saveLocation();
 
             ctx.writer->outputBeginNested(tagname, true);
             ctx.writer->outputBeginArray(item_tag.get());
@@ -724,7 +724,7 @@ void Esdl2Array::process(Esdl2TransformerContext &ctx, IPropertyTree *pt, const 
             {
                 ctx.writer->outputEndArray(item_tag.get());
                 ctx.writer->outputEndNested(tagname); //we need to close out the nested area first
-                ctx.writer->rewindTo(prevlen); //rewind
+                ctx.writer->rewindTo(prevLocation); //rewind
             }
             else
             {
@@ -773,7 +773,7 @@ void Esdl2Array::process(Esdl2TransformerContext &ctx, const char *out_name, Esd
     }
     else
     {
-        int prevlen = ctx.writer->length();
+        Owned<IInterface> prevLocation = ctx.writer->saveLocation();
         ctx.writer->outputBeginNested(xml_tag.get(), true);
         ctx.writer->outputBeginArray(item_tag.get());
         int curlen = ctx.writer->length();
@@ -801,7 +801,7 @@ void Esdl2Array::process(Esdl2TransformerContext &ctx, const char *out_name, Esd
         {
             ctx.writer->outputEndArray(item_tag.get());
             ctx.writer->outputEndNested(xml_tag.get()); // we need to close out this section first
-            ctx.writer->rewindTo(prevlen); //rewind
+            ctx.writer->rewindTo(prevLocation); //rewind
         }
         else
         {
@@ -878,7 +878,7 @@ void Esdl2Struct::process(Esdl2TransformerContext &ctx, IPropertyTree *pt, const
 
     if (checkVersion(ctx))
     {
-        unsigned prevlen = ctx.writer->length();
+        Owned<IInterface> prevLocation = ctx.writer->saveLocation();
         if (out_name && *out_name) {
             if (!might_skip_root || !ctx.skip_root)
                 ctx.writer->outputBeginNested(out_name, true);
@@ -907,7 +907,7 @@ void Esdl2Struct::process(Esdl2TransformerContext &ctx, IPropertyTree *pt, const
                 if (ctx.writer->length() == curlen) //nothing was added, empty content, remove open tag
                 {
                     ctx.writer->outputEndNested(out_name); //we need to close out current section first
-                    ctx.writer->rewindTo(prevlen); //rewind
+                    ctx.writer->rewindTo(prevLocation); //rewind
                 }
                 else
                     ctx.writer->outputEndNested(out_name);
@@ -924,8 +924,8 @@ void Esdl2Struct::process(Esdl2TransformerContext &ctx, const char *out_name, Es
         ctx.xppp->skipSubTree();
     else
     {
-        unsigned prevlen = ctx.writer->length();
-        unsigned curlen = prevlen;
+        Owned<IInterface> prevLocation = ctx.writer->saveLocation();
+        unsigned curlen = ctx.writer->length();
 
         if (out_name && *out_name)
         {
@@ -1007,6 +1007,7 @@ void Esdl2Struct::process(Esdl2TransformerContext &ctx, const char *out_name, Es
                             Esdl2Base& chd = *child;
 
                             unsigned len = ctx.writer->length();
+                            Owned<IInterface> location = ctx.writer->saveLocation();
                             local.dataForProcessed = false;
 
                             chd.process(ctx, NULL, &local);
@@ -1019,7 +1020,7 @@ void Esdl2Struct::process(Esdl2TransformerContext &ctx, const char *out_name, Es
                                 {
                                     ESDL_DBG("Taking out data for DataFor '%s' from out buffer", chd.queryDataFor()->queryName());
                                     local.setDataFor(chd.queryDataFor()->queryName(), ctx.writer->str()+len);
-                                    ctx.writer->rewindTo(len);
+                                    ctx.writer->rewindTo(location);
                                 }
                             }
 
@@ -1055,7 +1056,7 @@ void Esdl2Struct::process(Esdl2TransformerContext &ctx, const char *out_name, Es
                 if (ctx.writer->length() == curlen) //nothing was added, empty content, remove open tag
                 {
                     ctx.writer->outputEndNested(out_name); //we need to close out current section first
-                    ctx.writer->rewindTo(prevlen); //rewind
+                    ctx.writer->rewindTo(prevLocation); //rewind
                 }
                 else
                 {

--- a/plugins/javaembed/javaembed.cpp
+++ b/plugins/javaembed/javaembed.cpp
@@ -2081,9 +2081,10 @@ public:
     {
         throwUnexpected();
     }
-    virtual void rewindTo(unsigned int prevlen)
+    virtual IInterface *saveLocation() const {return nullptr;}
+    virtual void rewindTo(IInterface *loc)
     {
-        //needs to be a no-op because it is used, but the way its used to trim empty xml sections I think we're fairly safe.
+        //needs to be a no-op because it is  used, but the way its used to trim empty xml sections I think we're fairly safe.
         //revisit cleaning up any empty objects later.
     }
     inline IEsdlDefStruct *queryCurrentEsdlStruct()

